Marlin/src/feature/bedlevel/bdl/bdl.cpp

```diff
@@ -126,38 +137,57 @@ void BDS_Leveling::process() {
       if (BD_I2C_SENSOR.BD_Check_OddEven(tmp) == 0) SERIAL_ECHOLNPGM("errorCRC");
     #endif
 
-    if ((tmp & 0x3FF) > 1020) {
+    if ((tmp & 0x3FF) > 1000) {
       BD_I2C_SENSOR.BD_i2c_stop();
       safe_delay(10);
     }
-
+    // Read version and Usually used as a connection check
+    if (config_state == -1) {
+      BD_I2C_SENSOR.BD_i2c_write(CMD_READ_VERSION);
+      safe_delay(100);
+      char tmp_1[21];
+      for (int i = 0; i < 19; i++) {
+        tmp_1[i] = BD_I2C_SENSOR.BD_i2c_read() & 0xFF;
+        safe_delay(50);
+      }
+      config_state = 0;
+      BD_I2C_SENSOR.BD_i2c_write(CMD_END_READ_CALIBRATE_DATA);
+      SERIAL_ECHOLNPGM("BDsensor version:", tmp_1);
+      safe_delay(50);
+    }
     // read raw calibrate data
-    if (config_state == -5) {
+    else if (config_state == -5) {
       BD_I2C_SENSOR.BD_i2c_write(CMD_START_READ_CALIBRATE_DATA);
-      safe_delay(1000);
+      safe_delay(100);
 
       for (int i = 0; i < MAX_BD_HEIGHT * 10; i++) {
         tmp = BD_I2C_SENSOR.BD_i2c_read();
         SERIAL_ECHOLNPGM("Calibrate data:", i, ",", tmp & 0x3FF, ", check:", BD_I2C_SENSOR.BD_Check_OddEven(tmp));
-        safe_delay(500);
+        safe_delay(50);
       }
       config_state = 0;
       BD_I2C_SENSOR.BD_i2c_write(CMD_END_READ_CALIBRATE_DATA);
-      safe_delay(500);
+      safe_delay(50);
     }
     else if (config_state <= -6) {   // Start Calibrate
-      safe_delay(100);
+      safe_delay(10);
       if (config_state == -6) {
         //BD_I2C_SENSOR.BD_i2c_write(1019); // begin calibrate
         //delay(1000);
+        const millis_t old_inactive_time = gcode.stepper_inactive_time;
         gcode.stepper_inactive_time = SEC_TO_MS(60 * 5);
-        gcode.process_subcommands_now(F("M17 Z"));
-        gcode.process_subcommands_now(F("G1 Z0.0"));
-        z_pose = 0;
+        //gcode.process_subcommands_now(F("M17 Z"));
+        gcode.process_subcommands_now(F("G1Z-0.05"));
+        safe_delay(1000);
+        gcode.process_subcommands_now(F("G92Z0\nG1Z0.05\nG92Z0"));
         safe_delay(1000);
+        while (planner.get_axis_position_mm(Z_AXIS) > 0.00001f) safe_delay(1);
```

Comment:
`G92` is not the most ideal choice since it modifies the workspace offset. It doesn't set the actual current position per se. So the better choice here would be to set `current_position.z` directly and call `sync_plan_position`.

Reply:
Thank you very much! I have replaced the G92.